x/gamm/twap/api_test.go

```diff
@@ -0,0 +1,204 @@
+		},
+		"(1 record) start and end point to same record, end time = now": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTime.Add(time.Minute),
+			input:        makeSimpleTwapInput(baseTime, baseTime.Add(time.Minute), quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+
+		"(2 record) start and end point to same record": {
+			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			ctxTime:      baseTime.Add(time.Minute),
+			input:        makeSimpleTwapInput(baseTime, tPlusOne, quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(2 record) start and end exact, different records": {
+			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			ctxTime:      baseTime.Add(time.Minute),
+			input:        makeSimpleTwapInput(baseTime, baseTime.Add(10*time.Second), quoteAssetA),
+			expTwap:      sdk.NewDec(10),
+		},
+		"(2 record) start exact, end after second record": {
+			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			ctxTime:      baseTime.Add(time.Minute),
+			input:        makeSimpleTwapInput(baseTime, baseTime.Add(20*time.Second), quoteAssetA),
+			expTwap:      sdk.NewDecWithPrec(75, 1), // 10 for 10s, 5 for 10s
+		},
+		"(2 record) start exact, end after second record, sp1": {
+			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			ctxTime:      baseTime.Add(time.Minute),
+			input:        makeSimpleTwapInput(baseTime, baseTime.Add(20*time.Second), quoteAssetB),
+			expTwap:      sdk.NewDecWithPrec(15, 2), // .1 for 10s, .2 for 10s
+		},
+		"(2 record) start and end interpolated": {
+			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record},
+			ctxTime:      baseTime.Add(time.Minute),
+			input:        makeSimpleTwapInput(baseTime.Add(5*time.Second), baseTime.Add(20*time.Second), quoteAssetA),
+			// 10 for 5s, 5 for 10s = 100/15 = 6 + 2/3 = 6.66666666
+			expTwap: osmoutils.ThreePlusOneThird.MulInt64(2),
+		},
+
+		// error catching
+		"end time in future": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTime,
+			input:        makeSimpleTwapInput(baseTime, tPlusOne, quoteAssetA),
+			expErrorStr:  "future",
+		},
+		"start time after end time": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTime,
+			input:        makeSimpleTwapInput(tPlusOne, baseTime, quoteAssetA),
+			expErrorStr:  "after",
+		},
+		"start time too old (end time = now)": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTime,
+			input:        makeSimpleTwapInput(baseTime.Add(-time.Hour), baseTime, quoteAssetA),
+			expErrorStr:  "too old",
+		},
+		"start time too old": {
+			recordsToSet: []types.TwapRecord{baseRecord},
+			ctxTime:      baseTime.Add(time.Second),
+			input:        makeSimpleTwapInput(baseTime.Add(-time.Hour), baseTime, quoteAssetA),
+			expErrorStr:  "too old",
+		},
+		// TODO: overflow tests, multi-asset pool handling, make more record interpolation cases
```

Comment:
Agree that these would be useful. Do we have an issue for adding these?

Reply:
We do for multi-asset pool handling. Not yet for overflow, but handling overflow is one of the big remaining twap items